### PR TITLE
syncer(dm): enhance error handling about add inject at dml event

### LINKED ISSUE #4260: binlog inject command supports inject at dml
### Is your feature request related to a problem?

Now `binlog inject` command just can inject at ddl position, but user need inject at dml when handling error.

### Describe the feature you'd like

as title.

### PR DESCRIPTION
**Pull request by WizardXiao**
### What problem does this PR solve?
Issue Number: close #4260

### What is changed and how it works?
Add dml as the last event to injected events, syncer will execute injected events and then continue to sync binlog events. 

### Check List 

Tests 

 - Unit test
 - Integration test

Related changes

 - Need to update the documentation

### Release note 

```release-note
`binlog inject` command can inject ddls at dml.
```

### CODE REVIEW

File: dm/syncer/err-operator/operator.go [hunk 1]
```diff
@@ -34,11 +34,10 @@ import (
...
 // used by `handle-error`.
 type Operator struct {
-	uuid          string // add a UUID, make it more friendly to be traced in log
-	op            pb.ErrorOp
-	isAllInjected bool                       // is all injected, used by inject
-	events        []*replication.BinlogEvent // startLocation -> events
-	originReq     *pb.HandleWorkerErrorRequest
+	uuid      string // add a UUID, make it more friendly to be traced in log
+	op        pb.ErrorOp
+	events    []*replication.BinlogEvent // startLocation -> events
```
Comment (Ehco1996) [inline dm/syncer/err-operator/operator.go:39]:
```suggestion
	events    []*replication.BinlogEvent //  events sorted by startLocation
```

File: dm/syncer/err-operator/operator.go [hunk 2]
```diff
@@ -209,22 +192,26 @@ func (h *Holder) MatchAndApply(startLocation, endLocation binlog.Location, realE
...
+			// set the last replace event as end position
+			e := operator.events[len(operator.events)-1]
+			e.Header.EventSize = endLocation.Position.Pos - startLocation.Position.Pos
+			e.Header.LogPos = endLocation.Position.Pos
+			if e, ok := e.Event.(*replication.QueryEvent); ok {
+				if endLocation.GetGTID() != nil {
+					e.GSet = endLocation.GetGTID().Origin()
+				}
 			}
+		} else {
```
Comment (Ehco1996) [inline dm/syncer/err-operator/operator.go:213]: better use clear check (elseif pb.ErrorOp_Inject) because we have many other operations

File: dm/syncer/syncer.go [hunk 3]
```diff
@@ -1872,19 +1873,22 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
 			if !s.isReplacingOrInjectingErr {
-				apply, op := s.errOperatorHolder.MatchAndApply(startLocation, currentLocation, e.Header.Timestamp)
+				apply, op := s.errOperatorHolder.MatchAndApply(startLocation, currentLocation, e)
 				if apply {
 					if op == pb.ErrorOp_Replace || op == pb.ErrorOp_Inject {
 						s.isReplacingOrInjectingErr = true
 						// revert currentLocation to startLocation
 						currentLocation = startLocation
 					} else if op == pb.ErrorOp_Skip {
+						queryEvent, _ := ev.(*replication.QueryEvent)
```
Comment (Ehco1996) [inline dm/syncer/syncer.go:1891]:
```suggestion
						queryEvent := ev.(*replication.QueryEvent)
```
if we can make sure this event is queryEvent, other we should use ok pattern

File: dm/syncer/syncer.go [hunk 4]
```diff
@@ -1924,16 +1928,11 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
 				s.isReplacingOrInjectingErr = false
 				s.locations.reset(currentLocation)
-				if s.errOperatorHolder.IsInject(startLocation) {
-					s.errOperatorHolder.SetHasAllInjected(startLocation)
-					// reset event as startLocation, avoid to be marked in checkpoint
-					currentLocation.Position.Pos = startLocation.Position.Pos
-					err = s.streamerController.RedirectStreamer(tctx, startLocation)
-				} else {
+				if !s.errOperatorHolder.IsInject(startLocation) {
 					err = s.streamerController.RedirectStreamer(tctx, currentLocation)
```
Comment (Ehco1996) [inline dm/syncer/syncer.go:1932]:
```suggestion
					if err := s.streamerController.RedirectStreamer(tctx, currentLocation);err!=nil{
					return err
					}
```

**Review by Ehco1996 — COMMENTED** [review]
rest LGTM

File: dm/syncer/err-operator/operator.go (hunk 1 above)
Comment (WizardXiao) [reply dm/syncer/err-operator/operator.go:39]: `events` is constructed by ddls in `binlog inject ddl1, ddl2....`. it will add `LogPos ` and `EventSize ` when `MatchAndApply` and the `LogPos ` will be differnet because of `replace` or `inject`.

File: dm/syncer/err-operator/operator.go (hunk 2 above)
Comment (WizardXiao) [reply dm/syncer/err-operator/operator.go:213]: ok

File: dm/syncer/syncer.go (hunk 3 above)
Comment (WizardXiao) [reply dm/syncer/syncer.go:1891]: ok

File: dm/syncer/syncer.go (hunk 4 above)
Comment (WizardXiao) [reply dm/syncer/syncer.go:1932]: ok

File: dm/syncer/err-operator/operator.go (hunk 1 above)
Comment (Ehco1996) [reply dm/syncer/err-operator/operator.go:39]: ohh i see, this commnet `startLocation -> events` is misunderstood 😷

File: dm/syncer/syncer.go [hunk 5]
```diff
@@ -1948,16 +1952,10 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
-					// reset event as startLocation, avoid to be marked in checkpoint
-					currentLocation.Position.Pos = startLocation.Position.Pos
-					err = s.streamerController.RedirectStreamer(tctx, startLocation)
-				} else {
-					err = s.streamerController.RedirectStreamer(tctx, currentLocation)
-				}
-				if err != nil {
-					return err
+				if !s.errOperatorHolder.IsInject(startLocation) {
+					if err = s.streamerController.RedirectStreamer(tctx, currentLocation); err != nil {
```
Comment (Ehco1996) [inline dm/syncer/syncer.go:1956]: seems we can remove the outed commnet `also redirect stream to next event` in L1950, now only redirect stream when replace

File: dm/syncer/syncer.go [hunk 6]
```diff
@@ -1873,7 +1873,8 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
 		// we calculate startLocation and endLocation(currentLocation) for Query event here
 		// set startLocation empty for other events to avoid misuse
 		startLocation = binlog.Location{}
-		if ev, ok := e.Event.(*replication.QueryEvent); ok {
+		switch ev := e.Event.(type) {
```
Comment (Ehco1996) [inline dm/syncer/syncer.go:1876]: i think we can delete TODO in L1872 `TODO: support all event` 🎉

File: dm/syncer/err-operator/operator.go (hunk 1 above)
Comment (WizardXiao) [reply dm/syncer/err-operator/operator.go:39]: Maybe `ddls -> events` is better, because most of events is  constructed by ddls.

File: dm/syncer/syncer.go (hunk 5 above)
Comment (WizardXiao) [reply dm/syncer/syncer.go:1956]: ok

File: dm/syncer/syncer.go (hunk 6 above)
Comment (WizardXiao) [reply dm/syncer/syncer.go:1876]: i think so too, `QueryEvent` and `RowsEvent` are the major event type.

**Review by Ehco1996 — APPROVED** [review]
LGTM

**Comment by GMHDBJD** [comment]
Some corner cases need to be carefully considered. If a transaction with multiple dmls, which failed at the second DML in DM worker, the error position will be at the second dml, and inject DDL may have some problem.

**Comment by WizardXiao** [comment]
> 

ok, i think i should add more integration test about these corner cases.

**Review by lance6716 — APPROVED** [review]

**Review by GMHDBJD — CHANGES_REQUESTED** [review]
> i think i should add more integration test about these corner cases.

/hold

**Comment by lance6716** [comment]
ping @WizardXiao do you want to add more integration test now or in next PR?

**Comment by WizardXiao** [comment]
this pr is holded for a long time, i think i will add more integration test in next pr. please help me unhold, thank you. @lance6716
